smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/AuthSchemeResolverGenerator.kt

```diff
@@ -0,0 +1,254 @@
+import software.amazon.smithy.swift.codegen.utils.clientName
+import software.amazon.smithy.swift.codegen.utils.toLowerCamelCase
+
+class AuthSchemeResolverGenerator() {
+    fun render(ctx: ProtocolGenerator.GenerationContext) {
+        val rootNamespace = ctx.settings.moduleName
+        val serviceIndex = ServiceIndex(ctx.model)
+
+        ctx.delegator.useFileWriter("./$rootNamespace/${ClientRuntimeTypes.Core.AuthSchemeResolver.name}.swift") {
+            renderResolverParams(serviceIndex, ctx, it)
+            it.write("")
+            renderResolverProtocol(ctx, it)
+            it.write("")
+            renderDefaultResolver(serviceIndex, ctx, it)
+            it.write("")
+            it.addImport(SwiftDependency.CLIENT_RUNTIME.target)
+            it.addImport("AWSClientRuntime")
+        }
+    }
+
+    private fun renderResolverParams(
+        serviceIndex: ServiceIndex,
+        ctx: ProtocolGenerator.GenerationContext,
+        writer: SwiftWriter
+    ) {
+        writer.apply {
+            openBlock(
+                "public struct ${getSdkId(ctx)}${ClientRuntimeTypes.Core.AuthSchemeResolverParameters.name}: \$L {",
+                "}",
+                ServiceTypes.AuthSchemeResolverParams
+            ) {
+                write("public let operation: String")
+                // If service supports SigV4 auth scheme, it's a special-case
+                // Region has to be in params in addition to operation string from AuthSchemeResolver protocol
+                if (serviceIndex.getEffectiveAuthSchemes(ctx.service).contains(SigV4Trait.ID)) {
+                    write("// Region is used for SigV4 auth scheme")
+                    write("public let region: String?")
+                }
+            }
+        }
+    }
+
+    private fun renderResolverProtocol(ctx: ProtocolGenerator.GenerationContext, writer: SwiftWriter) {
+        writer.apply {
+            openBlock(
+                "public protocol ${getSdkId(ctx)}${ClientRuntimeTypes.Core.AuthSchemeResolver.name}: \$L {",
+                "}",
+                ServiceTypes.AuthSchemeResolver
+            ) {
+                // This is just a parent protocol that all auth scheme resolvers of a given service must conform to.
+                write("// Intentionally empty.")
+                write("// This is the parent protocol that all auth scheme resolver implementations of")
+                write("// the service ${getSdkId(ctx)} must conform to.")
+            }
+        }
+    }
+
+    private fun renderDefaultResolver(
+        serviceIndex: ServiceIndex,
+        ctx: ProtocolGenerator.GenerationContext,
+        writer: SwiftWriter
+    ) {
+        val sdkId = getSdkId(ctx)
+        val defaultResolverName = "Default$sdkId${ClientRuntimeTypes.Core.AuthSchemeResolver.name}"
+        val serviceProtocolName = sdkId + ClientRuntimeTypes.Core.AuthSchemeResolver.name
+
+        writer.apply {
+            writer.openBlock(
```

Comment:
since you did `writer.apply` you should be able to call `openBlock` directly

Reply:
Changed as pointed out.